lib/sshutils/scp/scp.go

```diff
@@ -532,8 +534,17 @@ func (cmd *command) receiveFile(st *state, fc newFileCmd, ch io.ReadWriter) erro
 func (cmd *command) receiveDir(st *state, fc newFileCmd, ch io.ReadWriter) error {
 	cmd.log.Debugf("scp.receiveDir(%v): %v", cmd.Flags.Target, fc.Name)
 
-	st.push(fc.Name, st.stat)
-	err := cmd.FileSystem.MkDir(st.path.join(), int(fc.Mode))
+	if cmd.FileSystem.IsDir(cmd.Flags.Target[0]) {
+		// Copying into an existing directory? append to it:
+		st.push(fc.Name, st.stat)
+	} else {
+		// If target specifies a new directory, we need to reset
+		// state with it
+		st.path = newPathFromDirAndTimes(cmd.Flags.Target[0], st.stat)
+	}
```

Comment:
This could cause unexpected data duplication, especially in automated scripts which may simply rely on the fact that `tsh scp` always overwrites its target. For example:

```bash
$ stat /tmp/zz
stat: cannot stat '/tmp/zz': No such file or directory
$ tsh scp -r ./large-tree/ /tmp/zz/
[...]
$ diff ./large-tree/ /tmp/zz/ && echo same tree
same tree
$ tsh scp -r ./large-tree/ /tmp/zz/   # re-run the same tsh scp command
[...]
$ diff ./large-tree/ /tmp/zz/
Only in /tmp/zz/: large-tree
```
The files of `large-tree` would end up in two copies under `/tmp/zz`: once directly under `/tmp/zz`, once under `/tmp/zz/large-tree`.

Reply:
Yes, this is consistent with OpenSSH client's behavior - initial attempt to copy a directory into a new non-existing location causes a rename, but existing location appends to it.

Reply:
I can't test it right now but IIRC, OpenSSH `scp` changes its dir-copying behaviour [based on the trailing slash](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=76054) which I think `tsh scp` would not.

Reply:
That could be possible - I'll test that as well.

Reply:
I did not see any difference between trailing slash and no slash in behavior.

Reply:
client (ubuntu):
```
$ ssh -V
OpenSSH_8.4p1, OpenSSL 1.1.1f  31 Mar 2020
```
server (centos):
```
$ ssh -V
OpenSSH_7.4p1, OpenSSL 1.0.2k-fips  26 Jan 2017
```
Trailing slash does not change the behavior as described in the linked issue.

Reply:
I'm getting the same results.

Reply:
Yes, the server is `sshd`. It does not have a version flag but outputs the version anyways:
```
$ sshd -V
unknown option -- V
OpenSSH_7.4p1, OpenSSL 1.0.2k-fips  26 Jan 2017
```